File: src/main/python/Day10.py
import Utils as Utils

CLOSE_CHAR = ( ')', ']', '}', '>' )
CLOSES = { '(': ')', '[': ']', '{': '}', '<': '>' }
FIRST_SCORE = { ')': 3, ']': 57, '}': 1197, '>': 25137 }
CLOSE_SCORE = { '(': 1, '[': 2, '{': 3, '<': 4 }

def getFirstScore(str):
    listStart = []
    for x in str:
        if x in CLOSE_CHAR:
            prev = listStart.pop()
            
            if CLOSES[prev] != x:
                return FIRST_SCORE[x]
        else:
            listStart.append(x)
    return 0

def getCloseScore(str):
    listStart = []
    for x in str:
        if x in CLOSE_CHAR:
            prev = listStart.pop()
            
            if CLOSES[prev] != x:
                return -1
        else:
            listStart.append(x)
    score = 0
    listStart.reverse()
    for x in listStart:
        score = score * 5 + CLOSE_SCORE[x]
    return score

@Utils.time_it_and_evaluate
def partOne(lines):
    score = 0
    for line in lines:
        score += getFirstScore(line)
    return score

@Utils.time_it_and_evaluate
def partTwo(lines):
    scores = []
    for line in lines:
        value = getCloseScore(line)
        if value > 0:
            scores.append(value)
    scores.sort()
    return scores[int(len(scores) / 2)]

if __name__ == '__main__':
    lines = Utils.readAllLines(2021, 'day10')
    
    print('Day 10')
    partOne(lines)
    partTwo(lines)

File: src/main/java/me/hardcoded/aoc2021/Day10.java
package me.hardcoded.aoc2021;

import java.util.*;

import me.hardcoded.util.DayBase;
import me.hardcoded.util.Utils;

public class Day10 extends DayBase {
	public static void main(String[] args) throws Exception {
		new Day10().run();
	}
	
	public Day10() {
		super(2021, 10);
	}
	
	public void run() throws Exception {
		List<String> lines = Utils.readAllLines(this);
		
		Utils.startPrintf(toString());
		Utils.printf("PartOne: %s\n", partOne(lines));
		Utils.printf("PartTwo: %s\n", partTwo(lines));
	}
	
	public static final Map<Character, Character> CLOSES = Map.of('(', ')', '[', ']', '{', '}', '<', '>');
	public static final Map<Character, Long> FIRST_SCORE = Map.of(')', 3L, ']', 57L, '}', 1197L, '>', 25137L);
	public static final Map<Character, Long> CLOSE_SCORE = Map.of('(', 1L, '[', 2L, '{', 3L, '<', 4L);
	public static final Set<Character> CLOSE_CHAR = Set.of(')', ']', '}', '>');
	
	public long getFirstScore(String str) {
		LinkedList<Character> stack = new LinkedList<>();
		
		for (int i = 0, len = str.length(); i < len; i++) {
			char next = str.charAt(i);
			
			if (CLOSE_CHAR.contains(next)) {
				char prev = stack.pollLast();
				char expected = CLOSES.get(prev);
				
				if (next != expected) {
					return FIRST_SCORE.get(next);
				}
			} else {
				stack.add(next);
			}
		}
		
		return 0;
	}
	
	public long getCloseScore(String str) {
		LinkedList<Character> stack = new LinkedList<>();
		
		for (int i = 0, len = str.length(); i < len; i++) {
			char next = str.charAt(i);
			
			if (CLOSE_CHAR.contains(next)) {
				char prev = stack.pollLast();
				char expected = CLOSES.get(prev);
				
				if (next != expected) {
					return -1;
				}
			} else {
				stack.add(next);
			}
		}
		
		long score = 0;
		for (int i = stack.size() - 1; i >= 0; i--) {
			score = score * 5 + CLOSE_SCORE.get(stack.get(i));
		}
		
		return score;
	}
	
	public long partOne(List<String> lines) throws Exception {
		long score = 0;
		for (String line : lines) {
			score += getFirstScore(line);
		}
		return score;
	}
	
	public long partTwo(List<String> lines) throws Exception {
		List<Long> scores = new ArrayList<>();
		for (String line : lines) {
			long value = getCloseScore(line);
			if (value > 0) {
				scores.add(value);
			}
		}
		scores.sort(null);
		return scores.get(scores.size() / 2);
	}
}
